src/iceberg/expression/residual_evaluator.cc

```diff
@@ -0,0 +1,565 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "iceberg/expression/residual_evaluator.h"
+
+#include "iceberg/expression/binder.h"
+#include "iceberg/expression/expression.h"
+#include "iceberg/expression/expression_visitor.h"
+#include "iceberg/expression/expressions.h"
+#include "iceberg/expression/predicate.h"
+#include "iceberg/partition_spec.h"
+#include "iceberg/row/struct_like.h"
+#include "iceberg/schema.h"
+#include "iceberg/schema_internal.h"
+#include "iceberg/transform.h"
+#include "iceberg/util/checked_cast.h"
+#include "iceberg/util/macros.h"
+
+namespace iceberg {
+
+class ResidualVisitor : public BoundVisitor<std::shared_ptr<Expression>> {
+ public:
+  ResidualVisitor(const std::shared_ptr<PartitionSpec>& spec,
+                  const std::shared_ptr<Schema>& schema, const StructLike& partition_data,
+                  bool case_sensitive)
+      : spec_(spec),
+        schema_(schema),
+        partition_data_(partition_data),
+        case_sensitive_(case_sensitive) {
+    ICEBERG_ASSIGN_OR_THROW(auto partition_type_, spec_->PartitionType(*schema));
+    partition_schema_ = FromStructType(std::move(*partition_type_), std::nullopt);
+  }
+
+  Result<std::shared_ptr<Expression>> AlwaysTrue() override {
+    return Expressions::AlwaysTrue();
```

Comment:
We cannot use `Expressions` family in this file because they may throw. Please use corresponding `Make` functions.